_posts/2019-02-19-refactoring-state-pattern.markdown

```diff
@@ -0,0 +1,114 @@
+---
+layout: post
+title:  "Refactoring with Design Patterns - The State Pattern"
+date: 2019-02-19 10:00:00
+categories: ["code-refactor", "design-patterns"]
+author: "cleiviane"
+---
+
+In this series of [code refactoring](https://www.ombulabs.com/blog/tags/code-refactor) posts we are discussing how design patterns can be used to make our Ruby code beautiful, maintainable and clean.
+
+Today I want to talk about a pattern that can be very useful when we need to control the flow of a set of events of our objects: *The State Pattern* or a.k.a *Finite State Machine*.
+
+As a developer it's something common to see objects changing of states. At the beginning manage the state of an object can be as simple as have some boolean attributes where you can check if the object is in state A or B. But when the complexity is increasing you can end up with a number of states that is difficulty to manage without break the [SOLID principles](http://butunclebob.com/ArticleS.UncleBob.PrinciplesOfOod). That is where we can benefit of the elegant solution the State Pattern give to us.
+
+<!--more-->
+
+## The State Pattern
+
+According to the [Refactoring Guru](https://refactoring.guru/design-patterns/state): <strong>_"The main idea of the State patter is that, at any given moment, there’s a finite number of states which a object can be in. Within any unique state, the program behaves differently, and the program can be switched from one state to another instantaneously._</strong>
+
+Think in the solution as a [finite sate machine](https://brilliant.org/wiki/finite-state-machines/) where you can control your objects making sure that it can be in only one state at the same time, so the objects will transit from one state to another to perform a set of different actions.
+
+Here at Ombulabs a good example of when we need to use state pattern it's in one of our products, the [Ombushop](https://secure.ombushop.com). It's an e-commerce platform where a user can create it's customized store and sell products online.
+
+One of the objects we need to handle is the `Order` (of course), which will change of state several times until we can say that the order is complete. The order if first created when the buyer adds a product to the cart, then if the buyer is in the checkout page the order will change to checkout state, when the payment is done we need to change the order to pending and that way the order will move from one state to other until it's completed or canceled. This is a very typical scenario where we can use the state.
```

Comment:
@cleicar `The order if` -> `The order is`

Reply:
`from one state to other` -> `from one state to another`

Reply:
@cleicar I think you missed this comment: https://github.com/ombulabs/blog/pull/175#discussion_r258193941